File: labs/daemon/modules/relay/multipath_dedup.go
package relay

import (
	"sync"
)

type MultipathDedupBuffer struct {
	mu             sync.Mutex
	expectedSeq    uint64
	seenHistory    map[uint64]struct{}
	reorderQueue   map[uint64][]byte
	maxHistorySize int
}

func NewMultipathDedupBuffer(initialSeq uint64, maxHistory int) *MultipathDedupBuffer {
	if maxHistory <= 0 {
		maxHistory = 128
	}
	return &MultipathDedupBuffer{
		expectedSeq:    initialSeq,
		seenHistory:    make(map[uint64]struct{}),
		reorderQueue:   make(map[uint64][]byte),
		maxHistorySize: maxHistory,
	}
}

func (b *MultipathDedupBuffer) Ingest(seq uint64, data []byte) [][]byte {
	b.mu.Lock()
	defer b.mu.Unlock()

	// Drop duplicate
	if seq < b.expectedSeq {
		return nil
	}
	if _, seen := b.seenHistory[seq]; seen {
		return nil
	}
	if _, exists := b.reorderQueue[seq]; exists {
		return nil
	}

	b.seenHistory[seq] = struct{}{}
	if len(b.seenHistory) > b.maxHistorySize {
		// Prune
		for k := range b.seenHistory {
			if k < b.expectedSeq {
				delete(b.seenHistory, k)
			}
		}
	}

	b.reorderQueue[seq] = data

	var ready [][]byte
	for {
		pkt, ok := b.reorderQueue[b.expectedSeq]
		if !ok {
			break
		}
		ready = append(ready, pkt)
		delete(b.reorderQueue, b.expectedSeq)
		b.expectedSeq++
	}

	return ready
}

func (b *MultipathDedupBuffer) ExpectedSeq() uint64 {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.expectedSeq
}

File: labs/control-ui/api-shadow/multipath_dedup.ts
/**
 * Multipath Packet Deduplication Buffer
 */
export class MultipathDedupBuffer {
  private seenHistory: Set<number> = new Set();
  private reorderQueue: Map<number, Uint8Array> = new Map();
  public expectedSeq: number;
  public maxHistorySize: number;
  constructor(expectedSeq: number = 1, maxHistorySize: number = 128) {
    this.expectedSeq = expectedSeq;
    this.maxHistorySize = maxHistorySize;
  }

  ingest(seq: number, data: Uint8Array): Uint8Array[] {
    if (seq < this.expectedSeq || this.seenHistory.has(seq) || this.reorderQueue.has(seq)) {
      return [];
    }

    this.seenHistory.add(seq);
    if (this.seenHistory.size > this.maxHistorySize) {
      const min = Math.min(...this.seenHistory);
      this.seenHistory.delete(min);
    }

    this.reorderQueue.set(seq, data);

    const ready: Uint8Array[] = [];
    while (this.reorderQueue.has(this.expectedSeq)) {
      ready.push(this.reorderQueue.get(this.expectedSeq)!);
      this.reorderQueue.delete(this.expectedSeq);
      this.expectedSeq++;
    }

    return ready;
  }
}
